src/csharp/Microsoft.Spark/ML/Util/Read.cs

```diff
@@ -0,0 +1,67 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Linq;
+using System.Reflection;
+using Microsoft.Spark.Interop.Ipc;
+using Microsoft.Spark.Sql;
+
+namespace Microsoft.Spark.ML.Feature
+{
+    /// <summary>
+    /// Class for utility classes that can load ML instances.
+    /// </summary>
+    /// <typeparam name="T">ML instance type</typeparam>
+    public class ScalaMLReader<T> : IJvmObjectReferenceProvider
+    {
+        public ScalaMLReader(JvmObjectReference jvmObject) => Reference = jvmObject;
+
+        public JvmObjectReference Reference { get; private set; }
+
+        /// <summary>
+        /// Loads the ML component from the input path.
+        /// </summary>
+        /// <param name="path">The path the previous instance of type T was saved to</param>
+        /// <returns>The type T instance</returns>
+        public T Load(string path) =>
+            WrapAsType((JvmObjectReference)Reference.Invoke("load", path));
+
+        /// <summary>Sets the Spark Session to use for saving/loading.</summary>
+        /// <param name="sparkSession">The Spark Session to be set</param>
+        public ScalaMLReader<T> Session(SparkSession sparkSession)
+        {
+            Reference.Invoke("session", sparkSession);
+            return this;
+        }
+
+        private static T WrapAsType(JvmObjectReference reference)
+        {
+            ConstructorInfo constructor = typeof(T)
+                .GetConstructors(BindingFlags.NonPublic | BindingFlags.Instance)
+                .Single(c =>
+                {
+                    ParameterInfo[] parameters = c.GetParameters();
+                    return (parameters.Length == 1) &&
+                        (parameters[0].ParameterType == typeof(JvmObjectReference));
+                });
+
+            return (T)constructor.Invoke(new object[] { reference });
+        }
+    }
+
+    /// <summary>
+    /// Interface for objects that provide MLReader.
+    /// </summary>
+    /// <typeparam name="T">
+    /// ML instance type
+    /// </typeparam>
+    public interface ScalaMLReadable<T>
```

Comment:
prepend interface with `I`

Reply:
separate file.

Reply:
May I know why we need to separate this into a new file? As we're actually just exposing IScalaMLReadable and ScalaMLReader is the real implementation of the read function.